### PR TITLE
Change the type of `Parallelism` field in BroadcastJob

### PR DESCRIPTION
**Pull request by xlgao-zju**
Signed-off-by: Xianglin Gao <xianglin.gxl@alibaba-inc.com>

### CODE REVIEW

File: pkg/apis/apps/v1alpha1/broadcastjob_types.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ type BroadcastJobSpec struct {
 	// be less than this number when the work left to do is less than max parallelism.
 	// Not setting this value means no limit.
 	// +optional
-	Parallelism *int32 `json:"parallelism,omitempty" protobuf:"varint,1,opt,name=parallelism"`
+	Parallelism *string `json:"parallelism,omitempty" protobuf:"varint,1,opt,name=parallelism"`
```
Comment (xiang90) [inline pkg/apis/apps/v1alpha1/broadcastjob_types.go]: is this a backward compatible change? if not, add this to change log.

File: pkg/controller/broadcastjob/broadcastjob_controller.go [hunk 2]
```diff
@@ -300,15 +304,24 @@ func addLabelToPodTemplate(job *appsv1alpha1.BroadcastJob) {
...
 	// max concurrent running pods
 	var parallelism int32
 	if job.Spec.Parallelism == nil {
 		// not specify,take the max int
 		parallelism = int32(1<<31 - 1)
 	} else {
-		parallelism = *job.Spec.Parallelism
+		parallelismIntStr := intstr.Parse(*job.Spec.Parallelism)
+		if parallelismIntStr.Type == intstr.String {
+			parallelismInt, err := strconv.Atoi(strings.TrimSuffix(parallelismIntStr.StrVal, "%"))
```
Comment (xiang90) [inline pkg/controller/broadcastjob/broadcastjob_controller.go]: move this to a util func? func percentageToAbsolute for example.

File: pkg/controller/broadcastjob/broadcastjob_controller_test.go [hunk 3]
```diff
@@ -54,7 +54,7 @@ func TestReconcileJobCreatePod(t *testing.T) {
 	_ = appsv1alpha1.AddToScheme(scheme)
 	_ = v1.AddToScheme(scheme)
 
-	p := int32(2)
+	p := "2"
```
Comment (xiang90) [inline pkg/controller/broadcastjob/broadcastjob_controller_test.go]: we should alao add a test for percentage value.

**Comment by xiang90** [comment]
let us separate this into 3 different PRs for the 3 new features.

File: docs/concepts/broadcastJob/README.md [hunk 4]
```diff
@@ -28,10 +28,10 @@ BroadcastJob.
 ### Parallelism
 
 `Parallelism` specifies the maximal desired number of Pods that should be run at
-any given time. By default, there's no limit.
+any given time. By default, there's no limit.`Parallelism` can be an int or a percent.
 
 For example, if a cluster has ten nodes and `Parallelism` is set to three, there can only be
-three pods running in parallel. A new Pod is created only after one running Pod finishes.
+three pods running in parallel. a cluster has ten nodes and `Parallelism` is set to 20%, there can only be two pods running in parallel. A new Pod is created only after one running Pod finishes.
```
Comment (Fei-Guo) [inline docs/concepts/broadcastJob/README.md]: How about this 

For example, if a cluster has ten nodes and `Parallelism` is set to three, there can only be
three pods running in parallel, or if a cluster has ten nodes and `Parallelism` is set to 20%, there can only be two pods running in parallel.

File: pkg/apis/apps/v1alpha1/broadcastjob_types.go (hunk 1 above)
Comment (Fei-Guo) [inline pkg/apis/apps/v1alpha1/broadcastjob_types.go]: Why not Parallelism intstr.IntOrString?  I think it is ok to break backward compatibility for now as long as we document it well in change log.

**Comment by xlgao-zju** [comment]
> let us separate this into 3 different PRs for the 3 new features.

`Add FailurePolicy in BroadcastJob` depends on ` Support Pause in BroadcastJob`, since we have a FailurePolicy called `Pause`. So, I will file two PRs.

**Comment by xlgao-zju** [comment]
@Fei-Guo updated, PTAL

**Pull request merged**